eth/beacon/state_machines/base.py

```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
+            active_state: ActiveState) -> Tuple[CrystallizedState, ActiveState]:
+        # TODO: it's a stub
+        return crystallized_state, active_state
+
+    #
+    #
+    # Proposer APIs
+    #
+    #
+    def propose_block(
+        self,
+        crystallized_state: CrystallizedState,
+        active_state: ActiveState,
+        block: BaseBeaconBlock,
+        shard_id: int,
+        shard_block_hash: Hash32,
+        chaindb: BaseBeaconBlock,
+        config: BeaconConfig,
+        private_key: int
+    ) -> Tuple[BaseBeaconBlock, CrystallizedState, ActiveState, 'AttestationRecord']:
+        """
+        Propose the given block.
+        """
+        block, post_crystallized_state, post_active_state = self.process_block(
+            crystallized_state,
+            active_state,
+            block,
+            chaindb,
+            config,
+        )
+
+        # Set state roots
+        block.copy(
+            crystallized_state_root=post_crystallized_state.hash,
+            active_state_root=post_active_state,
+        )
+
+        proposer_attestation = self.attest_proposed_block(
+            post_crystallized_state,
+            post_active_state,
+            block,
+            shard_id,
+            shard_block_hash,
+            chaindb,
+            config.CYCLE_LENGTH,
+            private_key,
+        )
+        return block, post_crystallized_state, post_active_state, proposer_attestation
+
+    def _update_the_states(self,
+                           crystallized_state: CrystallizedState,
+                           active_state: ActiveState) -> None:
+        self._crytallized_state = crystallized_state
+        self._active_state = active_state
+
+    def attest_proposed_block(self,
+                              post_crystallized_state: CrystallizedState,
+                              post_active_state: ActiveState,
+                              block: BaseBeaconBlock,
+                              shard_id: int,
+                              shard_block_hash: Hash32,
+                              chaindb: BaseBeaconBlock,
+                              cycle_length: int,
+                              private_key: int) -> 'AttestationRecord':
+        """
+        Return the initial attestation by the block proposer.
+
+        The proposer broadcasts zir attestation with the proposed block.
+        """
+        block_committees_info = get_block_committees_info(
+            block,
+            post_crystallized_state,
+            cycle_length,
+        )
+        # Vote
+        attester_bitfield = set_voted(
+            get_empty_bitfield(block_committees_info.proposer_committee_size),
+            block_committees_info.proposer_index_in_committee,
+        )
+
+        # Get justified_slot and justified_block_hash
+        justified_slot = post_crystallized_state.last_justified_slot
+        justified_block_hash = chaindb.get_canonical_block_by_slot(justified_slot).hash
```

Comment:
It may be nice to have both `get_canonical_block_hash_by_slot` and `get_canonical_block_by_slot` so that you can save the extra overhead of instantiating the block and then hashing it.  It's likely that the implementation already has to lookup the hash in order to lookup the block body.